src/waltz/xdp/fd_xdp1.c

```diff
@@ -32,56 +33,187 @@ struct __attribute__((aligned(8))) bpf_link_create {
   uint flags;
 };
 
+
+/*
+ 
+  ┌─────────────────────────────────────────┐
+  │   Packet Classification Control Flow    │
+  |                                         |
+  │       ┌────────┐                        │
+  │       │Ethernet│                        │
+  │       └───┬────┘                        │
+  │           │                             │
+  │       ┌───▼────┐                        │
+  │       │ IPV4   │                        │
+  │       └──┬───┬─┘                        │
+  │          │   │                          │
+  │          │   └────►──────┐              │
+  │          │        │ GRE  │              │
+  │          │        └───┬──┘              │
+  │          │            │                 │
+  │          │        ┌───▼────────┐        │
+  │          │        │ Inner IPV4 │        │
+  │          │        ├────────────┘        │
+  │          │        │                     │
+  │      ┌───▼──◄─────┘                     │
+  │      │ UDP  │                           │
+  │      └──┬───┘                           │
+  |         │                               |
+  |  ┌──────▼───────┐   ┌──────────────┐    |
+  |  | LBL_REDIRECT |   |   LBL_PASS   |    |
+  |  └──────────────┘   └──────────────┘    |
+  └─────────────────────────────────────────┘
+
+  fd_xdp_gen_program dynamically generates an eBPF bytecode program to 
+  classify incoming network packets in an XDP context. Each box in the above 
+  diagram represents a phase during the classification. NON-GRE packets follow 
+  the path of Ethernet-->IPV4-->UDP. GRE packets follow the path of 
+  Ethernet-->IPV4-->GRE-->Inner IPV4-->UDP. Based on the packet's protocol and 
+  specific header values, the function decides whether to redirect the packet 
+  to the firedancer net tile (LBL_REDIRECT) or pass the packet to the 
+  kernel (LBL_PASS).
+*/
 ulong
 fd_xdp_gen_program( ulong          code_buf[ 512 ],
                     int            xsks_fd,
                     uint           listen_ip4_addr,
                     ushort const * ports,
-                    ulong          ports_cnt ) {
+                    ulong          ports_cnt,
+                    bool           allowed_gre) {
+  
+  #define LBL_PASS      1   // Pass the packet to the kernel
+  #define LBL_REDIRECT  2   // Redirect the packet to firedancer software
+
+  #define LBL_GRE_CHECK     3   // Ethernet-->IPV4-->GRE-->Inner IPV4-->UDP
+  #define LBL_UDP_CHECK     4   // Ethernet-->IPV4-->UDP
```

Comment:
Just for sake of cleanliness, could you `#undef` these macros later in this file once you are done using them?